cpp/modmesh/universe/rtree.hpp

```diff
@@ -0,0 +1,700 @@
+#pragma once
+
+/*
+ * Copyright (c) 2025, An-Chi Liu <phy.tiger@gmail.com>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * - Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ * - Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ * - Neither the name of the copyright holder nor the names of its contributors
+ *   may be used to endorse or promote products derived from this software
+ *   without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+// This library implement R-tree based on "R-TREES" paper by Guttman in 1984
+
+#include <algorithm>
+#include <limits>
+#include <memory>
+#include <type_traits>
+#include <vector>
+
+namespace modmesh
+{
+
+/// Bounding box for 2D objects, e.g., Point2D, Segment2D, Triangle2D, etc.
+struct BoundingBox2D
+{
+    double min_x;
+    double min_y;
+    double max_x;
+    double max_y;
+
+    BoundingBox2D(double min_x_in, double min_y_in, double max_x_in, double max_y_in)
+        : min_x(min_x_in)
+        , min_y(min_y_in)
+        , max_x(max_x_in)
+        , max_y(max_y_in)
+    {
+    }
+
+    /// check if this bounding box intersects another bounding box
+    bool intersects(BoundingBox2D const & other) const
```

Comment:
Do not name it `intersects` because it is ambiguous.  Use `overlap`.  Do not use singular form of the verb.  Use base form.